lib/auth/recordingmetadata/recordingmetadatav1/thumbnails.go

```diff
@@ -0,0 +1,97 @@
+/**
+ * Teleport
+ * Copyright (C) 2025 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package recordingmetadatav1
+
+import (
+	"time"
+
+	"github.com/hinshun/vt10x"
+)
+
+// thumbnailBucketSampler maintains at most 'max' evenly-spaced thumbnails
+// from a stream of thumbnails without storing all of them in memory.
+// It dynamically adjusts which thumbnails to keep to maintain even spacing, doubling
+// the interval when the number of thumbnails exceeds 'max' and removing every second thumbnail
+// to keep the total count within 'max'.
+type thumbnailBucketSampler struct {
+	maxCapacity   int
+	entries       []*thumbnailEntry
+	interval      time.Duration
+	nextTimestamp time.Time
+	startTime     time.Time
+}
+
+type thumbnailEntry struct {
+	state       *vt10x.TerminalState
+	startOffset time.Duration
+	endOffset   time.Duration
+	timestamp   time.Time
+}
+
+func newThumbnailBucketSampler(maxCapacity int, interval time.Duration) *thumbnailBucketSampler {
+	return &thumbnailBucketSampler{
+		maxCapacity: maxCapacity,
+		entries:     make([]*thumbnailEntry, 0, maxCapacity),
+		interval:    interval,
+	}
+}
+
+func (s *thumbnailBucketSampler) shouldCapture(timestamp time.Time) bool {
+	return !timestamp.Before(s.nextTimestamp)
+}
+
+func (s *thumbnailBucketSampler) add(state *vt10x.TerminalState, timestamp time.Time) {
+	if s.startTime.IsZero() {
+		s.startTime = timestamp
+	}
+
+	if len(s.entries) >= s.maxCapacity {
+		s.adaptInterval()
+	}
+
+	entry := &thumbnailEntry{
+		state:       state,
+		timestamp:   timestamp,
+		startOffset: timestamp.Sub(s.startTime),
+		endOffset:   timestamp.Add(s.interval).Add(-1 * time.Millisecond).Sub(s.startTime),
```

Comment:
Nit: a short inline comment explaining why we subtract 1 millisecond may be valuable.